src/app-admin.js

```diff
@@ -0,0 +1,409 @@
+      reason: `${!userName ? 'user-cant-be-empty' : 'role-cant-be-empty'}`,
+      details: `${!userName ? 'Username' : 'Role'} cannot be empty`,
+    });
+    return;
+  }
+
+  const roleIdFromDb = UserService.validateRole(role);
+  if (!roleIdFromDb) {
+    res.status(400).send({
+      status: 'error',
+      reason: 'role-does-not-exists',
+      details: 'Selected role does not exist',
+    });
+    return;
+  }
+
+  const userIdInDb = UserService.getUserById(id);
+  if (!userIdInDb) {
+    res.status(400).send({
+      status: 'error',
+      reason: 'cannot-find-user-to-update',
+      details: `Cannot find user ${userName} to update`,
+    });
+    return;
+  }
+
+  UserService.updateUserWithRole(
+    userIdInDb,
+    userName,
+    displayName || null,
+    enabled ? 1 : 0,
+    role,
+  );
+
+  res.status(200).send({ status: 'ok', data: { id: userIdInDb } });
+});
+
+app.delete('/users', validateSessionMiddleware, async (req, res) => {
+  if (!isAdmin(res.locals.user_id)) {
+    res.status(403).send({
+      status: 'error',
+      reason: 'forbidden',
+      details: 'permission-not-found',
+    });
+    return;
+  }
+
+  const ids = req.body.ids;
+  let totalDeleted = 0;
+  ids.forEach((item) => {
+    const ownerId = UserService.getOwnerId();
+
+    if (item === ownerId) return;
+
+    UserService.deleteUserAccess(item);
+    UserService.transferAllFilesFromUser(ownerId, item);
+    const usersDeleted = UserService.deleteUser(item);
+    totalDeleted += usersDeleted;
+  });
+
+  if (ids.length === totalDeleted) {
+    res
+      .status(200)
+      .send({ status: 'ok', data: { someDeletionsFailed: false } });
+  } else {
+    res.status(400).send({
+      status: 'error',
+      reason: 'not-all-deleted',
+      details: '',
+    });
+  }
+});
+
+app.get('/access', validateSessionMiddleware, (req, res) => {
+  const fileId = req.query.fileId;
+
+  const { granted } = UserService.checkFilePermission(
+    fileId,
+    res.locals.user_id,
+  ) || {
+    granted: 0,
+  };
+
+  if (granted === 0 && !isAdmin(res.locals.user_id)) {
+    res.status(403).send({
+      status: 'error',
+      reason: 'forbidden',
+      details: 'permission-not-found',
+    });
+    return false;
+  }
+
+  const fileIdInDb = UserService.getFileById(fileId);
+  if (!fileIdInDb) {
+    res.status(404).send({
+      status: 'error',
+      reason: 'invalid-file-id',
+      details: 'File not found at server',
+    });
+    return false;
+  }
+
+  const accesses = UserService.getUserAccess(
+    fileId,
+    res.locals.user_id,
+    isAdmin(res.locals.user_id),
+  );
+
+  res.json(accesses);
+});
+
+app.post('/access', (req, res) => {
+  const userAccess = req.body || {};
+  const session = validateSession(req, res);
+
+  if (!session) return;
+
+  const { granted } = UserService.checkFilePermission(
+    userAccess.fileId,
+    session.user_id,
+  ) || {
+    granted: 0,
+  };
+
+  if (granted === 0 && !isAdmin(session.user_id)) {
+    res.status(400).send({
```

Comment:
the 404 is behind a forbidden 'wall', if the bad actor is a admin or a file owner, there is no leakage since they can already have access to file list